Support for specifying warehouse and role in snowflake URL string

abhinavcohesity opened the pull request:
@dhui There was an issue I came across with warehouse name and role not specified in URL string when trying to connect via code which I hadn't encountered earlier. Can you please approve ?

abhinavcohesity commented:
@dhui Any comments on this ? Otherwise we can merge. It's tested against hosted snowflake instance.

dhui commented on database/snowflake/README.md:
```diff
@@ -1,10 +1,14 @@
 # Snowflake
 
-`snowflake://user:password@accountname/schema/dbname?query`
+`snowflake://user:password@accountname/schema/dbname/warehouse?query`
```
warehouse should be a query param: https://godoc.org/github.com/snowflakedb/gosnowflake#hdr-Connection_Parameters

abhinavcohesity replied:
Done

dhui commented on database/snowflake/snowflake.go:
```diff
@@ -103,8 +105,8 @@ func (p *Snowflake) Open(url string) (database.Driver, error) {
-	if len(splitPath) < 3 {
-		return nil, ErrNoSchemaOrDatabase
+	if len(splitPath) < 4 {
+		return nil, ErrNoSchemaOrDatabaseOrWarehouse
```
warehouse shouldn't be required

abhinavcohesity replied:
Yes made it a query param.

dhui commented on database/snowflake/snowflake.go:
```diff
@@ -117,12 +119,24 @@ func (p *Snowflake) Open(url string) (database.Driver, error) {
+		return nil, ErrNoRole
+	}
+
 	cfg := &sf.Config{
```
It may make sense to:
1. Filter out migrate specific params from the connection URI
1. Rely on [sf.Config.ParseDSN](https://godoc.org/github.com/snowflakedb/gosnowflake#ParseDSN) to parse and validate the connection URI for you
1. Pass the filtered and validated connection URI to `sql.Open()`

abhinavcohesity replied:
Done.

dhui commented on database/snowflake/README.md:
```diff
@@ -1,10 +1,14 @@
+`snowflake://user:password@accountname/schema/dbname/warehouse?query`
+
+Example URL
+`snowflake://abhinav:abc@/ih12289.us-east-2.aws/public/UTIL_DB/TEST?x-migrations-table=schema_migrations&role=SYSADMIN`
```
Hope these aren't valid credentials...

abhinavcohesity replied:
Yes just dummy credentials.

abhinavcohesity commented:
@dhui Any more comments ? I added support so that snowflake driver can execute multiple queries in single migration file. Also removed the 'computeLineFromPos' function in snowflake.go, as it wasn't being hit anytime.

dhui commented on database/snowflake/snowflake.go:
```diff
@@ -177,65 +161,28 @@ func (p *Snowflake) Run(migration io.Reader) error {
+	// one go (https://support.snowflake.
+	// net/s/question/0D50Z00007ltRE7SAM?multiple-sql-statements-in-a-single
+	// -api-call-are-not-supported-use-one-api-call-per-statement-instead).
+	queries := strings.Split(queriesStr, ";")
```
Also, I've noticed that a few DB drivers implement this same logic so opened [another issue](https://github.com/golang-migrate/migrate/issues/406) to refactor this if you're interested in working on it. Thanks for all of your work so far!

abhinavcohesity replied:
@dhui I can create the util method and use it in snowflake driver. Other db drivers can use it.

dhui replied:
🙏

abhinavcohesity replied:
@dhui Reading the migration file line by line would need comment handling logic which might complicate things a little bit. For example
Comments of the following form can be there atleast for snowflake migrations:
/*
*
*
*/

CREATE TABLE;

-------------------------------------------

As of now by reading the whole migration file and splitting by ';' simplifies the code. Also as snowflake driver doesnt support executing multiple statements unlike Cassandra driver, should we add an option for user to specify this in the query param (x-multi-statement) ? Thoughts ?

dhui replied:
Did you encounter issues with [bufio.Reader.ReadSlice()](https://golang.org/pkg/bufio/#Reader.ReadSlice)? My understanding is that it should behave similarly to splitting the string without the memory overhead.

dhui replied:
Use [multistmt.Parse()](https://godoc.org/github.com/golang-migrate/migrate/database/multistmt#Parse)

driadi commented:
Any update on this PR?

ninthclowd commented:
I need this functionality for a project I'm working on in the next week.  If it's reasonable that this could get merged I would not have to fork and re-implement, but I understand there may be some outstanding issues that need to be resolved before this can go mainstream.  I would be happy to help contribute if it would get this feature into the main release.
